_posts/2016-10-17-test-stream.md

```diff
@@ -0,0 +1,299 @@
+---
+layout: post
+title:  "Testing Unbounded Pipeliens in Apache Beam"
+date:   2016-10-17 10:21:48 -0800
+excerpt_separator: <!--more-->
+categories: blog beam testing java
+authors:
+- tgroh
+---
+
+The Beam Programming Model unifies writing pipelines for Batch and Streaming
+pipelines. We’ve recently introduced a new PTransform to write tests for
+pipelines that will be run over unbounded datasets and must handle out-of-order
+and delayed data.
+<!--more-->
+Watermarks, Windows and Triggers form a core part of the Beam programming model
+-- they respectively determine how your data are grouped, when your input is
+complete, and when to produce results. This is true for all pipelines,
+regardless of if they are processing bounded or unbounded inputs. If you’re not
+familiar with watermarks, windowing, and triggering in the Beam model,
+[Streaming 101](https://www.oreilly.com/ideas/the-world-beyond-batch-streaming-101)
+and [Streaming 102](https://www.oreilly.com/ideas/the-world-beyond-batch-streaming-102)
+are an excellent place to get started. A key takeaway from
+these articles: in realistic streaming scenarios with intermittent failures and
+disconnected users, data can arrive out of order or be delayed. Beam’s
+primitives provide a way for users to perform useful, powerful, and correct
+computations in spite of these challenges.
+
+We need comprehensive tests that cover crucial failure scenarios and corner
```

Comment:
Perhaps preface with "As users of Beam, we..."  to clarify this isn't referring to testing Beam itself.